File: Leetcode practice/78. Subsets.cpp
#include <bits/stdc++.h>
using namespace std;


class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        int len = nums.size();
        int mask = (1<<len) - 1;

        vector<vector<int> > ret;
        vector<int> result;
        ret.push_back(result);

        while(mask > 0){

            result.clear();
            for(int i = 0;i<len;i++)
                if(mask & (1<<i)) result.push_back(nums[i]);
            ret.push_back(result);
            mask--;
        }
        return ret;

    }
};


int main()
{

}

File: go-leet-code/problem/78. Subsets.go
package problem

func subsets(nums []int) [][]int {
	ret := make([][]int, 0)
	var rec func(int, []int)
	rec = func(pos int, inter []int) {
		// fmt.Printf("%v %v\n", pos, inter)
		if pos == len(nums) {
			tmp := make([]int, len(inter))
			copy(tmp, inter)
			ret = append(ret, tmp)
			return
		}
		rec(pos+1, inter)
		inter = append(inter, nums[pos])
		rec(pos+1, inter)
		inter = inter[:len(inter)-1]
	}

	masking := func() [][]int {
		mask := (1 << uint(len(nums))) - 1
		for mask >= 0 {
			set := make([]int, 0)
			for i := 0; i < len(nums); i++ {
				if ((mask >> uint(i)) & 0x1) != 0 {
					set = append(set, nums[i])
				}
			}
			ret = append(ret, set)
			mask--
		}
		return ret
	}

	if len(nums)%2 == 1 {
		inter := make([]int, 0)
		rec(0, inter)
		return ret
	}
	return masking()

}
